File: 034/cpp/main.cpp
class Solution {
public:
    int searchRange1(vector<int>& nums, int target) {
        if (nums.empty()) {
            return -1;
        }
        if (nums[nums.size() - 1] < target) {
            return nums.size() - 1;
        }
        int i = -1;
        int j = nums.size() - 1;
        while (i + 1 < j) {
            int k = i + (j - i) / 2;
            if (nums[k] < target) {
                i = k;
            } else {
                j = k;
            }
        }
        return i;
    }

    int searchRange2(vector<int>& nums, int target) {
        if (nums.empty()) {
            return 0;
        }
        if (nums[0] > target) {
            return 0;
        }
        int i = 0;
        int j = nums.size();
        while (i + 1 < j) {
            int k = i + (j - i) / 2;
            if (nums[k] <= target) {
                i = k;
            } else {
                j = k;
            }
        }
        return j;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        int i = searchRange1(nums, target);
        int j = searchRange2(nums, target);
        if (i + 1 == j) {
            return {-1, -1};
        } else {
            return {i + 1, j - 1};
        }
    }
};

File: 034/rust/main.rs
impl Solution {
    fn search_range1(nums: &Vec<i32>, target: i32) -> i32 {
        let mut i = -1;
        let mut j = nums.len() as i32;
        while i + 1 < j {
            let k = (i + (j - i) / 2) as usize;
            if nums[k] < target {
                i = k as i32;
            } else {
                j = k as i32;
            }
        }
        return i;
    }

    fn search_range2(nums: &Vec<i32>, target: i32) -> i32 {
        let mut i = -1;
        let mut j = nums.len() as i32;
        while i + 1 < j {
            let k = (i + (j - i) / 2) as usize;
            if nums[k] <= target {
                i = k as i32;
            } else {
                j = k as i32;
            }
        }
        return j;
    }

    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let i = Solution::search_range1(&nums, target);
        let j = Solution::search_range2(&nums, target);
        if i + 1 == j {
            return vec![-1, -1];
        } else {
            return vec![i + 1, j -1];
        }
    }
}
